javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,126 @@ private[lang] object StackTrace {
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
+    }
     this
   }
 
+  def getCause(): Throwable = e
+
+  def getLocalizedMessage(): String = getMessage()
+
+  def getMessage(): String = s
+
   def getStackTrace(): Array[StackTraceElement] = {
-    if (stackTrace eq null) {
-      stackTrace = Array.empty
-    }
-    stackTrace
+    // Be robust! Test this.stackTrace against null rather than relying upon
+    // the value of writableStackTrace.
+    //
+    // subclass scala.util.control.NoStackTrace overrides fillInStackTrace()
+    // so that it never touches this.stackTrace. This creates the situation
+    // where writeableStackTrace is true and this.stackTrace is null.
+    //
+    // If scala code creates this situation, then by the Gell-Mann principle
+    // user code in the wild is bound to do so.
+    //
+    // If stackTrace is null, no profit to calling fillInStackTrace().
+    // If writableStackTrace is true, then fillInStackTrace has already
+    // been called in the constructor. If the stack is not writable, then
+    // it can not be filled in.
+
+    if (stackTrace == null) {
+      Array.empty[StackTraceElement] // as specified by Java 8.
+    } else
+      this.synchronized {
+        stackTrace.clone
+      }
   }
 
-  def setStackTrace(stackTrace: Array[StackTraceElement]): Unit = {
-    var i = 0
-    while (i < stackTrace.length) {
-      if (stackTrace(i) eq null)
-        throw new NullPointerException()
-      i += 1
+  final def getSuppressed(): Array[Throwable] = {
+    if (suppressed == null) {
+      Array.empty[Throwable]
+    } else
+      this.synchronized {
```

Comment:
The `synchronized` should also surround `if (suppressed == null)`, otherwise there can be an inconsistency.

Reply:
Changed as recommended. Thank you.